src/webgl/ShaderGenerator.js

```diff
@@ -1638,3 +1638,159 @@ export default shadergenerator;
+ * @property {{x: number, y: number}} texCoord - The texture coordinates (x, y) for the vertex.
+ * @property {{r: number, g: number, b: number, a: number}} color - The color at the vertex.
+ */
+
+/**
+ * @function getWorldInputs
+ * @experimental
+ * @description
+ * Registers a callback to modify the world-space properties of each vertex in a shader. This hook can be used inside {@link p5.baseMaterialShader}.modify() and similar shader modify calls to customize vertex positions, normals, texture coordinates, and colors before rendering. "World space" refers to the coordinate system of the 3D scene, before any camera or projection transformations are applied.
+ *
+ * This hook is available in:
+ * - {@link p5.baseMaterialShader}
+ * - {@link p5.baseNormalShader}
+ * - {@link p5.baseColorShader}
+ * - {@link p5.baseStrokeShader}
+ *
+ * @param {function(Vertex): Vertex} callback
+ *        A callback function which receives and returns a Vertex struct.
+ *
+ * @see {@link p5.baseMaterialShader}
+ * @see {@link p5.Shader#modify}
+ *
+ * @example
+ * <div modernizr='webgl'>
+ * <code>
+ * let myShader;
+ * function setup() {
+ *   createCanvas(200, 200, WEBGL);
+ *   myShader = baseMaterialShader().modify(() => {
+ *     getWorldInputs(inputs => {
+ *       // Move the vertex up and down in a wave
+ *       inputs.position.y += 20 * sin(
+ *         millis() * 0.001 + inputs.position.x * 0.05
+ *       );
+ *       return inputs;
+ *     });
+ *   });
+ * }
+ * function draw() {
+ *   background(255);
+ *   shader(myShader);
+ *   lights();
+ *   noStroke();
+ *   fill('red');
+ *   sphere(50);
+ * }
+ * </code>
+ * </div>
+ */
+
+/**
+ * @function combineColors
+ * @experimental
+ * @description
+ * Registers a callback to customize how color components are combined in the fragment shader. This hook can be used inside {@link p5.baseMaterialShader}.modify() and similar shader modify calls to control the final color output of a material. The callback receives a ColorComponents struct and should return an object with a `color` property ({ r, g, b }) and an `opacity` property (number).
+ *
+ * This hook is available in:
+ * - {@link p5.baseMaterialShader}
+ * - {@link p5.baseNormalShader}
+ * - {@link p5.baseColorShader}
+ * - {@link p5.baseStrokeShader}
+ *
+ * @param {function(ColorComponents): { color: {r: number, g: number, b: number}, opacity: number }} callback
+ *        A callback function which receives a ColorComponents struct and returns the final color and opacity.
+ *
+ * @see {@link p5.baseMaterialShader}
+ * @see {@link p5.Shader#modify}
+ *
+ * @example
+ * <div modernizr='webgl'>
+ * <code>
+ * let myShader;
+ * function setup() {
+ *   createCanvas(200, 200, WEBGL);
+ *   myShader = baseMaterialShader().modify(() => {
+ *     combineColors(components => {
+ *       // Custom color combination: add a red tint
+ *       let color = {
```

Comment:
Although you can *access* `.r`, `.g`, `.b`, and `.a` properties, the return value would need to be either `[r, g, b, a]` as an array, or more explicitly, `vec4(r, g, b, a)` -- we don't yet support an object notation for this. (Also it will need to be just the one object, and not an object with separate color and opacity properties, for it to compile correctly I think.)

Reply:
This also makes me wonder what the best way to document the `inputs` object is, since it's not exactly any normal type. @lukeplowden do you have any thoughts on this? We could call it a `vec4` and then explain in the description of the function what this means, or even not give it a type at all, and just have an explanation in the description.